### DISCUSSION TITLE
How to use widget in other widget? How to compose widget?

### DISCUSSION THREAD

**Discussion by linonetwo** [TiddlyWiki/TiddlyWiki5#6032] [Q&A]
I'm writing a widget, in `render()` I simply set innerHtml, but I hope I can render text using `$:/core/ui/ViewTemplate/body`

```ts
render(parent, nextSibling) {
      this.parentDomNode = parent;
      this.computeAttributes();

      const outputElement = this.document.createElement('article');
      outputElement.innerHTML = $tw.wiki.getTiddlerText(this.stateTiddlerTitle);
      parent.insertBefore(outputElement, nextSibling);
      this.domNodes.push(outputElement);
// how to push a $:/core/ui/ViewTemplate/body instead?
    }
```

It just output plain text...

<img width="765" alt="截屏2021-09-16 上午1 29 00" src="https://user-images.githubusercontent.com/3746270/133481110-2e4b43ba-6311-48f4-8579-1237443f893c.png">

I want to achieve something like this in JS:

```jsx
<$vars currentTiddler={{{[{!!title}addprefix[$:/state/linonetwo/zx-script/output/]] }}}>

{{||$:/core/ui/ViewTemplate/body}}

</$vars>
```

![图片](https://user-images.githubusercontent.com/3746270/133481159-5e44a596-97b6-4e7e-9fde-a6225f1d561f.png)

Is there an example to look at?

**Comment by saqimtiaz** [comment]
One approach is to create a parseTree for the children you want to create and insert that as the children for your widget in execute() and then call makeChildWidgets()

See https://github.com/Jermolene/TiddlyWiki5/discussions/5799#discussioncomment-870402

The other approach would be to use wiki.makeTranscludeWidget() method.
Example: https://github.com/Jermolene/TiddlyWiki5/blob/a409536ad01c46f7643588c75d8c819920f21778/core/modules/utils/dom/modal.js#L151

**Reply by linonetwo** [reply]
Sorry, but these examples still confused me. I may need some time to look at the source to get what these parameters mean.

I think I will use 

```js
render(parent, nextSibling) {
      const output = ($tw.wiki.getTiddlerText(this.stateTiddlerTitle) ?? '');
      const transcludeNode = $tw.wiki.makeTranscludeWidget('$:/core/ui/ViewTemplate/body', {
        type: "text",
        attributes: {
          text: {
            type: 'string',
            value: output
          }
        }
      });
      this.parseTreeNode.children.push(transcludeNode);

      this.makeChildWidgets([{
        attributes: this.parseTreeNode.attributes,
        children: this.parseTreeNode.children
      }]);
```

**Comment by linonetwo** [comment]
I was confused by the widget API, so I have to use wikitext to achieve transclusion...

```jsx
creator: LinOnetwo
tags: $:/tags/ViewTemplate
title: $:/plugins/linonetwo/zx-script/OutputBar
list-before: $:/plugins/linonetwo/inverse-link-and-folder/infobar
type: text/vnd.tiddlywiki

<$vars currentTiddler={{{[{!!title}addprefix[$:/state/linonetwo/zx-script/output/]is[missing]then[]]~[{!!title}addprefix[$:/state/linonetwo/zx-script/output/]] }}}>

{{||$:/core/ui/ViewTemplate/body}}

</$vars>
```
